### PR TITLE
Added T4 simple generator as a proof of concept with ReactAction as PR implementation

### PR DESCRIPTION
**Pull request by Muchaszewski**
The idea is to have a stable generator for all kinds of types to improve the readability of the code. I have used T4 in the past with great success! and the code is usually more readable then string builders.

As a proof of concept and as a content of this PR I have added `ReactAction` as an Action that can be used as Action in normal C# code, but have an `AddListener` method that can be used in react, 

C# example as a normal delegate
```
public ReactAction<string> del;
public void Foo()
{
    del += (val) => Debug.Log(val):
    del.Invoke("It worked");
}
```

but also can be used as a wrapper around Callback class

React
```
useEffect(() => {
    const unsubscribe = Global.Instance.del.AddListener((val) => {
        console.log("It worked from react");
    };
    return () => unsubscribe();
},[]);
```

### CODE REVIEW

File: Editor/Generators/RuntimeT4Generator.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        /// <summary>
+        ///     Additional data that can be passed to the templates. First key is the template name, value is Session
+        ///     data that is accessible from the template.
+        /// </summary>
+        public static Dictionary<string, Dictionary<string, object>> Sessions { get; } = new Dictionary<string, Dictionary<string, object>>();
+
+        /// <summary>
+        ///     Generates the code from all templates in the project and saves them to Runtime/Generated folder
+        /// </summary>
+        [MenuItem("React/GenerateCodeFromTemplates")]
```
Comment (KurtGokhan) [inline Editor/Generators/RuntimeT4Generator.cs:24]: Should be wrapped in `#if REACT_UNITY_DEVELOPER`

File: Editor/Generators/RuntimeT4Generator.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using ReactUnity.Editor.Editor.Generators.ReactAction;
+using UnityEditor;
+using UnityEngine;
+
+namespace ReactUnity.Editor.Editor.Generators
```
Comment (KurtGokhan) [inline Editor/Generators/RuntimeT4Generator.cs:11]: Namespace should have one `Editor`

File: Runtime/Generated/ReactAction.cs [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+            Action listener = null;
+            listener = () =>
+            {
+                try
+                {
+                    cb.Call();
+                }
+                catch
+                {
+                    _delegate -= listener;
```
Comment (KurtGokhan) [inline Runtime/Generated/ReactAction.cs:41]: It is not good to remove listener on `catch`. If this is about the refresh/hot-reload issue, there should be another way to fix this. For example, `Callback` can be marked as disposed when context is disposed. I will check what is viable.

File: Runtime/Generated/ReactAction.cs [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+// </auto-generated>
+// ------------------------------------------------------------------------------>
+using System;
+using ReactUnity.Helpers;
+
+namespace ReactUnity.Helpers
+{
+  /// <summary>
+  ///   A wrapper class for the <see cref="ReactUnity.Helpers.Callback" /> class to provide convienient access to the
+  ///   to delegates and events in ReactUnity.
```
Comment (KurtGokhan) [inline Runtime/Generated/ReactAction.cs:16]: Comment can be improved.

File: Runtime/Generated/ReactAction.cs [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+  {
+      /// <summary>
+      ///   The underlying action.
+      /// </summary>
+      private Action _delegate { get; set; }
+
+      /// <summary>
+      ///   Adds the listener to the action, and returns a function that removes the listener. This method should be only called from React.
+      /// </summary>
+      [Obsolete("This method should be only called from React.", true)]
```
Comment (KurtGokhan) [inline Runtime/Generated/ReactAction.cs:28]: I think this should only be a warning. Also, the message should discourage the usage from C#, but not completely prohibit it.

**Review by KurtGokhan — CHANGES_REQUESTED** [review]

File: Runtime/Generated/ReactAction.cs (hunk 5 above)
Comment (Muchaszewski) [reply Runtime/Generated/ReactAction.cs:28]: I wanted to keep the standard syntax of the delegates as much as possible with += and -=. That's why I wanted to discourage AddListener usage from C#.

If JS engine can access private methods, we can change this to internal to remove the Obsolete warning/error

File: Editor/Generators/ReactAction/ReactAction.Generator.tt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  /// <summary>
+  ///   A wrapper for the <see cref="ReactUnity.Helpers.Callback" /> class to provide convienient access to the
+  ///   delegates and events in ReactUnity from both JavaScript and TypeScript as well as C# as normal events.
+  /// </summary>
+	/// <example>
+	///   <code language="csharp">
+	///     var callback = new ReactAction&lt;string&gt;();
+	///     callback += (str) => Console.WriteLine(str);
+	///     callback.Invoke("It works! From C#");
+	///     callback -= (str) => Console.WriteLine(str);
```
Comment (KurtGokhan) [inline Editor/Generators/ReactAction/ReactAction.Generator.tt:28]: This example is wrong. Subtracted delegate is not the same reference as added delegate.

File: Runtime/Generated/ReactAction.cs (hunk 5 above)
Comment (KurtGokhan) [reply Runtime/Generated/ReactAction.cs]: Yes but `AddListener` returns a handy "RemoveListener" method, which could be useful also for C# users. `+` and `-` operators are prone to errors as you did with the example.

**Review by KurtGokhan — CHANGES_REQUESTED** [review]

File: Editor/Generators/ReactAction/ReactAction.Generator.tt (hunk 6 above)
Comment (Muchaszewski) [reply Editor/Generators/ReactAction/ReactAction.Generator.tt:28]: You are right, I didn't look closely what copilot suggested

File: Runtime/Generated/ReactAction.cs (hunk 5 above)
Comment (Muchaszewski) [reply Runtime/Generated/ReactAction.cs]: Before I push this, do you think we should have then 2 methods with the same signature?
```
public Action AddListener(object callback) // For React with Obsolete
public Action AddListener(Action callback) // For C#
```

or should we merge them into one, and resolve inside of it if it's Action or Callback?

Comment (KurtGokhan) [reply Runtime/Generated/ReactAction.cs]: I think another signature for C# is a good idea.

Comment (KurtGokhan) [reply Runtime/Generated/ReactAction.cs]: I think it should also be `AddListener` because we already use `AddListener` in several places.

**Comment by KurtGokhan** [comment]
Hey @Muchaszewski , I started fixing some issues now. If you have pending changes for this PR, please push them. I will soon merge this to main branch and do the necessary changes on it.

Also I added `Destroyed` property to callbacks. This may be the correct way to detect disposed callbacks.

**Pull request merged**